### PR TITLE
[portable-snippets] Fix array length issue

### PR DESCRIPTION
**Pull request by jimwang118**
```
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(34): error C2057: expected constant expression
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(34): error C2466: cannot allocate an array of constant size 0
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(372): error C2057: expected constant expression
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(372): error C2466: cannot allocate an array of constant size 0
```
The function `psnip_random_bytes` in the header file `random.h` defines the array `data[PSNIP_RANDOM_ARRAY_PARAM(length)]`. However, the size of this array is a variable. C language requires that the size of the array be a constant during compilation, so the compilation fails.
So the incoming array will be modified by the incoming pointer.

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] ~~SHA512s are updated for each updated download~~
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

Usage test pass with following triplets:

```
x86-windows
x64-windows
x64-windows-static
```

### CODE REVIEW

**Review by Cheney-W — DISMISSED** [review]

**Comment by jimwang118** [comment]
This problem has been submitted to the upstream issue [48](https://github.com/nemequ/portable-snippets/issues/48) for feedback, but there was no reply from the upstream for a long time, so I tried to fix it in vcpkg.

**Comment by BillyONeal** [comment]
> The function `psnip_random_bytes` in the header file `random.h` defines the array `data[PSNIP_RANDOM_ARRAY_PARAM(length)]`. However, the size of this array is a variable. C language requires that the size of the array be a constant during compilation, so the compilation fails.

This is not true: Variable Length Arrays (VLAs) exist. GCC implements them and I believe Clang implements them for compatibility with GCC, MSVC does not implement them. My understanding is that that feature is widely considered to be a mistake, but it's been in there since 1999.

Are there necessary ci.baseline.txt changes or something? (Why is this not causing the current builds to fail?)

File: ports/portable-snippets/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,7 @@ project(portable-snippets LANGUAGES C)
 include(GNUInstallDirs)
 
 option(PSNIP_INSTALL_HEADERS "Install header files" ON)
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c11" )
```
Comment (dg0yt) [inline ports/portable-snippets/CMakeLists.txt:8]: `CMAKE_C_STANDARD`?

File: ports/portable-snippets/fix_array.patch [hunk 2]
```diff
@@ -0,0 +1,89 @@
+diff --git a/random/random.c b/random/random.c
+index 0a83d00..a0d628c 100644
+--- a/random/random.c
++++ b/random/random.c
+@@ -31,7 +31,7 @@
+ #  endif
+ #endif
+ 
+-static int (* psnip_random_secure_generate)(size_t length, psnip_uint8_t data[PSNIP_RANDOM_ARRAY_PARAM(length)]) = NULL;
++static int (* psnip_random_secure_generate)(size_t length, psnip_uint8_t *data) = NULL;
```
Comment (dg0yt) [inline ports/portable-snippets/fix_array.patch:10]: The `PSNIP_RANDOM_ARRAY_PARAM` is defined in the header. Can't it choose a supported definition?
The default definition us empty, i.e.
~~~
static int (* psnip_random_secure_generate)(size_t length, psnip_uint8_t data[]) = NULL;
~~~

**Comment by jimwang118** [comment]
> > The function `psnip_random_bytes` in the header file `random.h` defines the array `data[PSNIP_RANDOM_ARRAY_PARAM(length)]`. However, the size of this array is a variable. C language requires that the size of the array be a constant during compilation, so the compilation fails.
> 
> This is not true: Variable Length Arrays (VLAs) exist. GCC implements them and I believe Clang implements them for compatibility with GCC, MSVC does not implement them. My understanding is that that feature is widely considered to be a mistake, but it's been in there since 1999.
> 
> Are there necessary ci.baseline.txt changes or something? (Why is this not causing the current builds to fail?)

This is an issue of VCPKG testing MSVC.
The issue with VCPKG testing MSVC is as follows: 
```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(101): error C2054: expected '(' to follow '_Noreturn'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(101): error C2085: 'thrd_exit': not in formal parameter list
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(104): error C2085: 'thrd_join': not in formal parameter list
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(105): error C2085: '_thrd_sleep32': not in formal parameter list
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(107): error C2085: '_thrd_sleep64': not in formal parameter list
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(116): error C2085: 'thrd_sleep': not in formal parameter list
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.36.32532\include\threads.h(116): error C2143: syntax error: missing ';' before '{'
```
This problem is due to the lack of compilation option -std:c11, after adding this option, a new problem has appeared.
```
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(34): error C2057: expected constant expression
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(34): error C2466: cannot allocate an array of constant size 0
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(372): error C2057: expected constant expression
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(372): error C2466: cannot allocate an array of constant size 0
```

File: ports/portable-snippets/fix_array.patch (hunk 2 above)
Comment (jimwang118) [reply ports/portable-snippets/fix_array.patch]: After adding the C11 option, the default options were not entered and a compilation error occurred.

**Comment by BillyONeal** [comment]
> This is an issue of VCPKG testing MSVC.

My point is, don't there need to be ci.baseline.txt or similar changes? If MSVC hates it why aren't we seeing it failing in CI builds?

**Comment by BillyONeal** [comment]
https://dev.azure.com/vcpkg/public/_build/results?buildId=97842

```console
Installing 1740/2229 portable-snippets:x64-windows@2019-09-20#3...
Building portable-snippets:x64-windows@2019-09-20#3...
-- Note: portable-snippets only supports static library linkage. Building static library.
-- Downloading https://github.com/nemequ/portable-snippets/archive/26496acb37ab46ee249ea19d45381da6955d89c4.tar.gz -> nemequ-portable-snippets-26496acb37ab46ee249ea19d45381da6955d89c4.tar.gz...
-- Extracting source D:/downloads/nemequ-portable-snippets-26496acb37ab46ee249ea19d45381da6955d89c4.tar.gz
-- Using source at D:/b/portable-snippets/src/a6955d89c4-b026d6ec08.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Performing post-build validation
Stored binaries in 1 destinations in 94.7 ms.
Elapsed time to handle portable-snippets:x64-windows: 2.1 s
```

Is the frontend team aware that they regressed this? I don't object to merging this but this seems like the kind of regression our testing is supposed to catch, not paper over.

**Comment by BillyONeal** [comment]
```
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(34): error C2057: expected constant expression
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(34): error C2466: cannot allocate an array of constant size 0
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(372): error C2057: expected constant expression
vcpkg\buildtrees\portable-snippets\src\a6955d89c4-b026d6ec08\random\random.c(372): error C2466: cannot allocate an array of constant size 0
```

This looks like a compiler bug to me. `PSNIP_RANDOM_ARRAY_PARAM` is defined as

```c++
#if defined(HEDLEY_ARRAY_PARAM)
#  define PSNIP_RANDOM_ARRAY_PARAM(expr) HEDLEY_ARRAY_PARAM(expr)
#elif defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L) && !defined(__cplusplus) && !defined(__PGI)
#  define PSNIP_RANDOM_ARRAY_PARAM(expr) (expr)
#else
#  define PSNIP_RANDOM_ARRAY_PARAM(expr)
#endif
```

which means that without `-std:c11`, `PSNIP_RANDOM_ARRAY_PARAM(length)` should expand to nothing. That means:

```c++
static int (* psnip_random_secure_generate)(size_t length, psnip_uint8_t data[PSNIP_RANDOM_ARRAY_PARAM(length)]) = NULL;
```

should expand to

```c++
static int (* psnip_random_secure_generate)(size_t length, psnip_uint8_t data[]) = NULL;
```

which is valid. The compiler appears wrong to treat a macro expanding to no tokens as a constant zero.

`-std:c11` changes `__STDC_VERSION__` to be greater than `199901L`, which changes the expansion to just `(length)` which is why you end up needing the subsequent patch.

Has a compiler frontend dev seen this?

File: ports/portable-snippets/fix_array.patch (hunk 2 above)
Comment (dg0yt) [reply ports/portable-snippets/fix_array.patch]: Note that you could simply add a single `&& !defined(_MSC_VER)` to the condition before the second definition alternative for `PSNIP_RANDOM_ARRAY_PARAM`. Like it already excludes `__PGI`.

**Comment by jimwang118** [comment]
> Has a compiler frontend dev seen this?

There is a related issue [1250747](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1250747) with compiler development, but this issue has not been fixed yet.

**Comment by BillyONeal** [comment]
> > Has a compiler frontend dev seen this?
> 
> There is a related issue [1250747](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1250747) with compiler development, but this issue has not been fixed yet. Do we need to wait for the compiler developers to fix this?

That seems unrelated since there is no `static` here.

**Comment by BillyONeal** [comment]
I believe the correct fix is https://github.com/nemequ/portable-snippets/pull/50

**Comment by jimwang118** [comment]
> I believe the correct fix is [nemequ/portable-snippets#50](https://github.com/nemequ/portable-snippets/pull/50)

OK, I'll fix it and test it.

**Comment by jimwang118** [comment]
Duplicate of https://github.com/microsoft/vcpkg/pull/36143.

**Pull request closed without merging**